main/src/main/java/org/mobilitydata/gtfsvalidator/validator/DuplicateFareRuleZoneIdFieldsValidator.java

```diff
@@ -41,27 +42,35 @@
 
   @Override
   public void validate(NoticeContainer noticeContainer) {
-    final Map<String, GtfsFareRule> fareRuleByZoneIdFieldsCombination =
+    final Map<Integer, GtfsFareRule> fareRuleByZoneIdFieldsCombination =
         new HashMap<>(fareRuleTable.entityCount());
     fareRuleTable
         .getEntities()
         .forEach(
             fareRule -> {
-              String fieldsCombination =
-                  fareRule.routeId()
-                      + fareRule.originId()
-                      + fareRule.containsId()
-                      + fareRule.destinationId();
-              if (fareRuleByZoneIdFieldsCombination.containsKey(fieldsCombination)) {
+              int fieldsCombinationHash = getZoneFieldsCombinationHash(fareRule);
+              if (fareRuleByZoneIdFieldsCombination.containsKey(fieldsCombinationHash)) {
```

Comment:
Similar to https://github.com/MobilityData/gtfs-validator/pull/806#discussion_r595456725, if you get the value from the HashMap and assign to a local variable and check for null instead of using `containsKey()`, you'll avoid multiple lookups here for the same hash.

Generally speaking, you only want to use `containsKey()` if you don't need to reference the value. If you need to reference the value, you can avoid the duplicate lookup by using `.get()` and checking for `null` instead.

Reply:
Fixed in https://github.com/MobilityData/gtfs-validator/pull/811/commits/553ba6d945455276dbbc2a44a8dbee9baf0aa534